README.rst
#########
pastemacs
#########

http://pypi.python.org/pypi/pastemacs/

**pastemacs** allows to create and retrieve pastes from
http://paste.pocoo.org directly from Emacs_.  It is available under the
terms of the `GNU GPL 2`_ (see ``COPYING``).

.. _Emacs: http://www.gnu.org/software/emacs/emacs.html
.. _`GNU GPL 2`: http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt


Installation
============

You will need at least Python 2.6 (earlier versions may work, but are not
tested, Python 3 is *not* supported).  pastemacs itself is available in the
`Python package index`_ and can simply be installed using ``pip`` or
``easy_install``::

   pip install pastemacs
   # or
   easy_install pastemacs

This installs pastemacs and its dependency lodgeitlib_. Alternatively you
can just clone the Git_ repository and install manually::

   git clone git://github.com/lunaryorn/pastemacs.git
   cd pastemacs
   python setup.py install

In this case you need to manually install lodgeitlib_ before.

Moreover you need a working installation of Pymacs_.  This should be
available in the package repositories of your distribution.  If not, or if
you are on Windows, refer to the `Pymacs documentation`_ for installation
instructions.

.. _`Python package index`: http://pypi.python.org/pypi/pastemacs
.. _lodgeitlib: http://packages.python.org/lodgeitlib
.. _git: http://git-scm.com/
.. _pymacs: http://pymacs.progiciels-bpi.ca/
.. _`Pymacs documentation`: http://pymacs.progiciels-bpi.ca/pymacs.html


Usage
=====

In order to use pastemacs, just add the following two lines to your Emacs
configuration file::

   (pymacs-load "pastemacs" "paste-")
   (paste-menu)

The second line only enables the GUI menu to create and fetch pastes, and is
not strictly required.  Now all pastemacs commands are available.  You can
use them either through the menu (if enabled) or by executing them directly
using ``M-x``.  pastemacs does not register any shortcuts for its commands,
so if you want some, you have to register them for yourself (e.g. by using
``global-set-key``).

Take a look at the configuration_, the author of pastemacs uses, if you want
to see an example.

.. _configuration:  http://github.com/lunaryorn/emacs-site/blob/master/50pasting.el


Creating pastes
---------------

The command ``paste-new`` creates a new paste.  If transient mark mode is
enabled and the mark is active, it pastes the contents of the region,
otherwise it pastes the complete buffer contents.

The command prompts for a (programming) language, in which the paste is
written.  This is used for syntax highlighting in the pastebin web
interface.  The list of available languages is fetched from the server, once
you create your first paste, and available as completion list for the
language prompt.  You cannot enter a language, which isn't supported, so if
the language isn't available, choose the default "text".

.. note::

   pastemacs tries to derive a language from the name of the major mode and
   sets it as default while prompting for a language, but this isn't really
   clever, and will often fail.  So be sure to check the language, if you
   want syntax highlighting for your paste.

By default the URL of the new paste will only be displayed in the minibuffer
and in the ``*Messages*`` buffer.  To automatically put the url into the
kill ring, set the variable ``paste-kill-url`` to a non-nil value in your
emacs configuration.

Alternatively new pastes can automatically be opened in your favourite web
browser by setting ``paste-show-in-browser`` to a non-nil value in your
emacs configuration.

Both variables are available in customization.


Retrieving pastes
-----------------

To retrieve a paste, use ``paste-fetch``.  It fetches a paste and inserts
its contents at the point in the current buffer.

The commands prompts for a paste to retrieve.  You can either enter a
complete paste URL, or just a paste ID.  Empty input is also allowed, and
stands for the last paste.


Configuration
-------------

Simply visit the customization group ``pastebin``.


Issues, Feedback and Development
================================

All development happens on Github_.

Please report issues, proposals or enhancements to the `issue tracker`_.
Please provide as much information as possible when reporting issues.

Feel free to clone the repository, add you own modifications and send a pull
requests.  Enhancements and new features are always appreciated :)

.. _github: https://github.com/lunaryorn/pastemacs
.. _`issue tracker`: https://github.com/lunaryorn/pastemacs/issues


CHANGES.rst
0.3 (Sep 12, 2010)
==================

- Pastes can be retrieved by URL now


0.2.1 (Sep 11, 2010)
====================

- Fixed a NameError
- ``paste-kill-url`` is on by default now


0.2 (Sep 10, 2010)
==================

- Merged ``paste-new-from-buffer`` and ``paste-new-from-region`` into a new,
  more convenient ``paste-new`` function.
- Removed ``paste-fetch-*`` functions.  There is no longer any function,
  which automatically creates a new buffer for a paste.  Do this for
  yourself, if required.
- Merged ``paste-insert-last`` and ``paste-insert-by-id`` into
  ``paste-fetch``.


0.1.2 (Aug 03, 2010)
====================

Fixed installation


0.1.1 (Nov 01, 2008)
====================

Initial release



pastemacs.py
# -*- coding: utf-8 -*-
# allows pasting the current buffer on paste.pocoo.org
# Copyright (c) 2007-2010 Sebastian Wiesner <lunaryorn@googlemail.com>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.


"""
    pastemacs
    =========

    This pymacs module provides a handful of function for convenient access
    to the pastebin at `http://paste.pocoo.org`.

    .. moduleauthor::  Sebastian Wiesner  <lunaryorn@googlemail.com>
"""


from Pymacs import lisp

from lodgeitlib import lodgeit


# necessary references
lisp.require(lisp['easymenu'])
lisp.require(lisp['browse-url'])


# customisation group
lisp("""
(defgroup pastebin nil
  "Access to the pastebin on paste.pocoo.org"
  :group 'convenience)

(defcustom paste-kill-url t
  "*If non-nil, put the url of a new paste into kill ring"
  :group 'pastebin
  :type 'boolean)

(defcustom paste-show-in-browser nil
  "*If non-nil, invoke the browser with the paste url after
  pasting a new snippet"
  :group 'pastebin
  :type 'boolean)
""")


def languages():
    """Returns a list of supported languages."""
    if not lodgeit.has_languages:
        lisp.message('Fetching list of supported languages from server')
    return lodgeit.languages.keys()


def read_language():
    """Reads a paste language from minibuffer. Provides completion based on
    the list ov available languages"""
    # guess language from major mode
    major_mode = lisp.major_mode.value().text
    def_language = major_mode[:-5]
    if def_language not in languages():
        def_language = 'text'
    msg = 'Language of paste snippet ({0}): '.format(def_language)
    language = (lisp.completing_read(msg, languages()).strip()
                or def_language)
    return language


def read_paste_id():
    """
    Read and return a paste ID or URL from the minibuffer.  An empty input
    is accepted, and supposed to stand for the last paste.
    """
    return lisp.read_no_blanks_input(
        'A paste ID or URL [default: last paste]: ')


def fetch(paste_id_or_url=None):
    """
    Fetch a paste and insert its content at point.

    If ``paste_id_or_url`` is given, fetch the paste with the given ID or
    URL, otherwise fetch the last paste.

    When called interactively, prompt for a paste ID or URL.  Empty input
    stands for the last paste.
    """
    if paste_id_or_url:
        paste = lodgeit.get_paste_by_id(paste_id_or_url)
    else:
        paste = lodgeit.get_last_paste()
    if paste:
        lisp.insert(paste.code)
    else:
        lisp.error('There is no paste {0}'.format(paste_id_or_url))
fetch.interaction = lambda: [read_paste_id()]


def new(language, region_start=None, region_end=None):
    """
    Create a new paste.  Use the given (programming) ``language`` for server
    side highlighting.

    If ``region_start`` and ``region_end`` are given, create a paste with
    the contents of this region.

    When called interactively with transient mark mode enabled and an active
    mark, create a paste with the contents of the region.  Otherwise create
    a paste with the contents of the whole buffer.
    """
    mark_active = lisp.mark_active.value()
    transient_mark_mode = lisp.transient_mark_mode.value()
    if lisp.interactive and  transient_mark_mode and mark_active:
        # use a region, if we have one
        region_start = lisp.region_beginning()
        region_end = lisp.region_end()
    elif region_start:
        # otherwise use the given arguments
        region_start = min(region_start, region_end)
        region_end = max(region_start, region_end)
    else:
        # as last resort, paste the whole buffer
        region_start = lisp.point_min_marker()
        region_end = lisp.point_max_marker()

    code = unicode(lisp.buffer_substring(region_start, region_end))
    filename = lisp.buffer_file_name()

    lisp.message('Transferring paste to server...')
    paste_id = lodgeit.new_paste(code, language, filename=filename)
    paste = lodgeit.get_paste_by_id(paste_id)
    lisp.message(
        'New paste with ID {0.id} created. Refer to {0.url}'.format(paste))
    if lisp.paste_kill_url.value():
        lisp.kill_new(paste.url)
    if lisp.paste_show_in_browser.value():
        lisp.browse_url(paste.url)
new.interaction = lambda: [read_language()]


def menu():
    """Creates a global menu to access the pastebin functionality"""
    lisp.easy_menu_add_item(
        lisp.current_global_map(),
        ['menu-bar'],
        ['Pastebin',
         ('Fetch...', lisp.paste_fetch),
         '---',
         ('New...', lisp.paste_new),
         ])
menu.interaction = ''


paste.py
# -*- coding: utf-8 -*-
# allows pasting the current buffer on paste.pocoo.org
# Copyright (c) 2007, 2008 Sebastian Wiesner <basti.wiesner@gmx.net>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.


"""
    paste
    =====

    This pymacs module provides a handful of function for convenient access
    to the pastebin at `http://paste.pocoo.org`.


    :copyright: 2008 by Sebastian Wiesner
    :license: GPL-2
"""


# TODO: add proper error handling to avoid exception tracebacks
# TODO: provide customisation to map pastebin languages to major modes


from xmlrpclib import ServerProxy

from Pymacs import lisp

from lodgeitlib import lodgeit


# necessary references
lisp.require(lisp['easymenu'])
lisp.require(lisp['browse-url'])


# customisation group
# XXX: maybe this could somehow be written in python
lisp("""
(defgroup pastebin nil
  "Access to the pastebin on paste.pocoo.org"
  :group 'convenience)

(defcustom paste-kill-url nil
  "*If non-nil, put the url of a new paste into kill ring"
  :group 'pastebin
  :type 'boolean)

(defcustom paste-show-in-browser nil
  "*If non-nil, invoke the browser with the paste url after
  pasting a new snippet"
  :group 'pastebin
  :type 'boolean)
""")


## lodgeit wrappers which add some messaging
def new_paste(code, language, filename):
    """Creates a new paste."""
    lisp.message('Transferring paste to server...')
    id = lodgeit.new_paste(code, language, filename=filename)
    paste = lodgeit.get_paste_by_id(id)
    lisp.message('New paste with id %s created. Refer to %s',
                 paste.id, paste.url)
    if lisp.paste_kill_url.value():
        lisp.kill_new(paste.url)
    if lisp.paste_show_in_browser.value():
        lisp.browse_url(paste.url)


def languages():
    """Returns a list of supported languages."""
    if not lodgeit.has_languages:
        lisp.message('Fetching list of supported languages from server')
    return lodgeit.languages.keys()


## NON-INTERACTIVE FUNCTIONS
# interactive argument completion
def read_language():
    """Reads a paste language from minibuffer. Provides completion based on
    the list ov available languages"""
    # guess language from major mode
    major_mode = lisp.major_mode.value().text
    def_language = major_mode[:-5]
    if def_language not in languages():
        def_language = 'text'
    msg = 'Language of paste snippet (%s): ' % def_language
    language = (lisp.completing_read(msg, languages()).strip()
                or def_language)
    return language


def get_new_from_region_args():
    """Gets args for new_from_region:
    The language and the currently active region"""
    lang = read_language()
    start, end = lisp.region_beginning(), lisp.region_end()
    return [start, end, lang]


def get_new_from_buffer_args():
    """Gets args for new_from_buffer:
    The paste language"""
    buff = lisp.read_buffer('Buffer to paste: ',
                            lisp.current_buffer(), True)
    lang = read_language()
    return [buff, lang]


def new_buffer_from_paste(paste):
    """Creates a new buffer from `paste`"""
    lisp.switch_to_buffer('paste %s' % paste.id)
    lisp.erase_buffer()
    lisp.insert(paste.code)
    # simple guessing of the buffer mode
    # XXX: is there a better way?
    # FIXME: do at least a bit of error handling and check, if there is such a mode
    mode = lisp['%s-mode' % paste.language]
    mode()


## INTERACTIVE FUNCTIONS
# to fetch pastes
def fetch_by_id(paste_id):
    """Fetches paste with `paste_id` and inserts it into a new buffer"""
    paste = lodgeit.get_paste_by_id(paste_id)
    if paste:
        new_buffer_from_paste(paste)
    else:
        lisp.error('There is no paste with id %s', paste_id)
fetch_by_id.interaction = ('nThe paste id: ')


def insert_by_id(paste_id):
    """Fetches paste with `paste_id` and inserts it into current buffer"""
    paste = lodgeit.get_paste_by_id(paste_id)
    if paste:
        lisp.insert(paste.code)
    else:
        lisp.error('There is no paste with id %s', paste_id)
insert_by_id.interaction = ('*nThe paste id: ')


def fetch_last():
    """Fetches last paste and inserts it into a new buffer"""
    paste = lodgeit.get_last_paste()
    new_buffer_from_paste(paste)
fetch_last.interaction = ''


def insert_last():
    """Inserts the last paste into current buffer"""
    paste = lodgeit.get_last_paste()
    lisp.insert(paste.code)
insert_last.interaction = '*'


# to create new pastes
def new_from_region(start, end, language):
    """Pastes the current selection"""
    code = unicode(lisp.buffer_substring(start, end))
    filename = lisp.buffer_file_name()
    new_paste(code, language, filename=filename)
new_from_region.interaction = get_new_from_region_args

def new_from_buffer(buffer, language):
    """Pastes the contents of buffer"""
    lisp.set_buffer(buffer)
    code = unicode(lisp.buffer_string())
    #code = lisp.buffer_substring(lisp.point_min(), lisp.point_max())
    filename = lisp.buffer_file_name()
    new_paste(code, language, filename=filename)
new_from_buffer.interaction = get_new_from_buffer_args


def menu():
    """Creates a global menu to access the pastebin functionality"""
    lisp.easy_menu_add_item(
        lisp.current_global_map(),
        ['menu-bar'],
        ['Pastebin',
         ('Fetch last', lisp.paste_fetch_last),
         ('Insert last', lisp.paste_insert_last),
         ('Fetch by id', lisp.paste_fetch_by_id),
         ('Insert by id', lisp.paste_insert_by_id),
         '---',
         ('New from buffer', lisp.paste_new_from_buffer),
         ('New from region', lisp.paste_new_from_region)
         ])
menu.interaction = ''
